src/Authenticator/ChangePasswordHandler.php

```
@@ -0,0 +1,149 @@
+            return $this->jsonResponse(
+                array_merge($schema, [
+                    'endpoints' => [
+                        'login' => $this->Link('mfa/login/{urlSegment}'),
+                        'complete' => $this->Link(),
+                    ],
+                    'shouldRedirect' => false,
+                ])
+            );
+        } catch (Throwable $exception) {
+            // If we don't have a valid member we shouldn't be here...
+            return $this->redirectBack();
+        }
+    }
+
+    public function mfa()
+    {
+        $store = $this->getStore();
+        if (!$store || !$store->getMember()) {
+            return $this->redirectBack();
+        }
+
+        $this->applyRequirements();
+
+        return [
+            'Form' => $this->renderWith($this->getViewerTemplates()),
+            'ClassName' => 'mfa',
+        ];
+    }
+
+    public function startMFACheck(HTTPRequest $request)
+    {
+        $store = $this->getStore();
+        $member = $store->getMember();
+
+        // If we don't have a valid member we shouldn't be here...
+        if (!$member) {
+            return $this->jsonResponse(['message' => 'Forbidden'], 403);
+        }
+
+        // Use the provided trait method for handling login
+        $response = $this->createStartLoginResponse(
+            $store,
+            Injector::inst()->get(MethodRegistry::class)->getMethodByURLSegment($request->param('Method'))
+        );
+
+        // Ensure detail is saved to the store
+        $store->save($request);
+
+        // Respond with our method
+        return $response;
+    }
+
+    public function verifyMFACheck(HTTPRequest $request)
+    {
+        $store = $this->getStore();
+
+        try {
+            $result = $this->verifyLoginRequest($store, $request);
+        } catch (InvalidMethodException $e) {
+            // Invalid method usually means a timeout. A user might be trying to verify before "starting"
```

Again - do you think it might be worth logging these exceptions instead of swallowing them?

---

Logging added